extensions/VisualEditor/wikia/modules/ve/dm/ve.dm.WikiaTransclusionModel.js

```diff
@@ -2,148 +2,188 @@
+					specs[page.title].params[page.params[i]] = {};
+				}
 			}
+			ve.extendObject( this.specCache, specs );
+		}
+	};
+
+	/**
+	 * @param {Object} specs
+	 * @param {Object} node
+	 * @param {string} sourceName
+	 * @param {Object} sourceMetadata
+	 */
+	function addInfoboxSourceParam( specs, node, sourceName, sourceMetadata ) {
+		// Single source can be used in multiple nodes.
+		// We override its spec with every usage, so only the last one stays.
+		// `data` node is the most generic one and doesn't give us any value.
+		// Let's ignore it if the source has already a meaningful type.
+		if ( specs.params[sourceName] && specs.params[sourceName].type !== 'data' ) {
+			return;
 		}
-		ve.extendObject( this.specCache, specs );
-	}
-};
 
-/**
- * @desc process the infobox params received from API. For all requested pages, if they contain infoboxes,
- * extend their param array with received infobox params.
- */
-ve.dm.WikiaTransclusionModel.prototype.fetchInfoboxParamsRequestDone = function ( specs, data ) {
-	var id, page, i, j, denormalizedData, source;
+		specs.params[sourceName] = {
+			label: sourceMetadata.label
+		};
 
-	if ( data && Object.keys( data.query ).length && Object.keys( data.query.pages ).length ) {
-		denormalizedData = this.denormalizeInfoboxTemplateTitles( data );
+		if ( sourceMetadata.primary === true ) {
+			specs.params[sourceName].type = node.type;
+		}
 
-		for ( id in denormalizedData ) {
-			page = denormalizedData[id];
+		specs.paramOrder.push( sourceName );
+	}
 
-			if ( !page.infoboxes ) {
-				return;
-			}
+	/**
+	 * @param {Object} node
+	 * @param {Object} specs
+	 */
+	function parseInfoboxMetadata( node, specs ) {
+		if ( node.type === 'group' ) {
+			node.metadata.forEach( function ( childNode ) {
+				parseInfoboxMetadata( childNode, specs );
+			} );
+		} else {
+			Object.keys( node.sources ).map( function ( sourceName ) {
+				var sourceMetadata = node.sources[sourceName];
+				addInfoboxSourceParam( specs, node, sourceName, sourceMetadata );
+			} );
+		}
+	}
 
-			if ( !specs[page.title] ) {
-				specs[page.title] = {
-					title: page.title,
-					description: '',
-					params: {},
-					paramOrder: []
-				};
-			}
+	/**
+	 * @desc process the infobox params received from API. For all requested pages, if they contain infoboxes,
+	 * extend their param array with received infobox params.
+	 */
+	ve.dm.WikiaTransclusionModel.prototype.fetchInfoboxParamsRequestDone = function ( specs, data ) {
+		var id, page, denormalizedData;
 
-			for ( i = 0; i < page.infoboxes.length; i++ ) {
-				page.infoboxes[i].sources = Object.keys( page.infoboxes[i].sourcelabels );
+		if ( data && Object.keys( data.query ).length && Object.keys( data.query.pages ).length ) {
+			denormalizedData = this.denormalizeInfoboxTemplateTitles( data );
 
-				for ( j = 0; j < page.infoboxes[i].sources.length; j++ ) {
-					source = page.infoboxes[i].sources[j];
+			for ( id in denormalizedData ) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Restrict what this loop acts on by testing each property. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AForIn)